src/hotspot/share/gc/g1/g1Policy.cpp

```diff
@@ -1400,6 +1404,85 @@ void G1Policy::calculate_optional_collection_set_regions(G1CollectionSetCandidat
                             num_optional_regions, max_optional_regions, prediction_ms);
 }
 
+// Number of regions required to store the given number of bytes, taking
+// into account the target amount of wasted space in PLABs.
+static size_t get_num_regions_adjust_for_plab_waste(size_t byte_count) {
+  size_t byte_count_adjusted = byte_count * (size_t)(100 + TargetPLABWastePct) / 100.0;
+
+  // Round up the region count
+  return (byte_count_adjusted + HeapRegion::GrainBytes - 1) / HeapRegion::GrainBytes;
+}
+
+bool G1Policy::proactive_collection_required(uint alloc_region_count) {
+  if (!Universe::is_fully_initialized()) {
+    // Don't attempt any proactive GC's before initialization is complete.
+    return false;
+  }
+
+  if (_g1h->young_regions_count() == 0 && _collection_set->candidates() != NULL && _collection_set->candidates()->is_empty()) {
```

Comment:
Thomas and I talked about this before but decided not to do it because the two cases where we're doing this are doing slightly different things (one is checking for NULL || empty and the other is checking !NULL && empty, which is not the inverse of the first). Looking at it closer now though, I think this one should also be bailing in the case where candidates is NULL right? If we do that, we can definitely add the has_candidates function and use that in both cases.